Support for Per Thread Default Stream (PTDS)

pentschev opened the pull request:
This PR adds Per Thread Default Stream (PTDS) support to CuPy, entirely controllable at runtime via an environment variable. The advantage of this approach is that multiple builds aren't necessary, greatly simplifying the shipping procedure of CuPy.

I don't currently know if this is the "right" approach for CuPy, but is certainly useful for us to get some experience with PTDS. Right now I've verified this with some pure Python multithreaded code (see below) and Dask, I was able to get a 2x speedup for some Dask workflows but I didn't do any performance testing with standalone CuPy as this is going to vary greatly w.r.t. application, number of threads, etc. As of now, I also don't know how to write tests to ensure that we aren't using the default stream when setting `CUPY_CUDA_PER_THREAD_DEFAULT_STREAM=1`, something I verified with the help of `nsys` and `nvprof`.

The following code serves as an example, verifiable with `nsys`/`nvprof`, each host thread runs on a different stream when `CUPY_CUDA_PER_THREAD_DEFAULT_STREAM=1` when compared to running only on the default stream (shown as stream `7`) when the variable isn't set:

```python
import threading

import cupy

NumThreads = 4


def thread_func():
    a = cupy.random.random((5000, 5000))
    c = cupy.fft.fft(a)


threads = [threading.Thread(target=thread_func, name="Thread %d" % (i,)) for i in range(NumThreads)]
for t in threads:
    t.start()
for t in threads:
    t.join()
```

Note we have a similar effort being done for PTDS support in RMM https://github.com/rapidsai/rmm/pull/633.

pentschev commented:
cc @leofang @quasiben @jakirkham

leofang reviewed (commented):
Just a quick pass. I have two questions:
1. Why does the CUDA Programming Guide say we need to recompile an app with `--default-stream`, while there exists such a simple solution?!
2. Why moving to PTDS will bring speedup in some of your tests? Is it because the implicit syncs are removed? If so, wouldn't using non-blocking streams also do the job for you? I am very curious! 😛

Thanks for pinging me btw!

leofang commented:
> As of now, I also don't know how to write tests to ensure that we aren't using the default stream when setting `CUPY_CUDA_PER_THREAD_DEFAULT_STREAM=1`, something I verified with the help of `nsys` and `nvprof`.

btw This was my impression too, but I just begin to wonder if it's possible to add a callback to the stream, and let it throw out something that we can capture in Python? CuPy already warps the stream callback API:
https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.Stream.html#cupy.cuda.Stream.add_callback
(Disclaimer: I've never tried this, I just read it from the Programming Guide 😅)

pentschev commented:
> 1. Why does the CUDA Programming Guide say we need to recompile an app with `--default-stream`, while there exists such a simple solution?!

Where does it say users _need_ to recompile? In https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html both alternatives are presented. The advantage of compiling with `--default-stream=per-thread` is that you need no code changes whatsoever when you're using the default stream already, but that may be incompatible when interfacing with applications that do not compile the same way.

> 2. Why moving to PTDS will bring speedup in some of your tests? Is it because the implicit syncs are removed? If so, wouldn't using non-blocking streams also do the job for you? I am very curious! 😛

They would, but using non-blocking streams can be _very_ challenging for RAPIDS, and Dask more specifically, we don't want users to require doing so on their own, so PTDS may be a very convenient solution, although it's still unclear whether we will indeed see major improvements for the average workflow. Below is a Dask example with 1 stream (i.e., thread) vs 4 streams:

![1 Dask worker with 1 thread](https://user-images.githubusercontent.com/4398246/100221125-890d4180-2f18-11eb-99eb-7841798f87e6.png)
![1 Dask worker with 4 threads](https://user-images.githubusercontent.com/4398246/100221135-8ad70500-2f18-11eb-9563-0a670ff5e260.png)

pentschev commented on cupy_backends/cuda/api/runtime.pxd:
```diff
@@ -686,6 +686,12 @@ cpdef enum:
 
+    # cudaStream
+    cudaStreamDefault = 0
+    cudaStreamNonBlocking = 1
```
Yeah, I wasn't sure if we wanted to have all in one place so I added that just in case but I'll remove it then.

pentschev replied:
Done.

pentschev commented on cupy/cuda/stream.pyx:
```diff
@@ -323,3 +333,4 @@ class ExternalStream(BaseStream):
 
 
 Stream.null = Stream(null=True)
+Stream.ptds = Stream(ptds=True)
```
That would work too, but I'm not sure if it's the right solution for two reasons:

1. The null stream may become ambiguous, are we really using the null stream or the per-thread one?
1. It's not anymore possible to enforce an explicit usage of the default stream when the environment variable is on, not sure how useful it is to use the default stream in that case anyway but it's probably useful to have that option.

leofang replied:
Yeah it's best for the core devs to decide, as I could be overlooking some design issues. My thoughts are:
1. `Stream.null` in CuPy means the default stream, and previously it meant the legacy one.
2. If a user explicitly asks for PTDS by setting the env var, it implicitly suggests that the legacy default stream should be voided to the maximum possible extent.

pentschev replied:
I agree that the legacy stream should be avoided when PTDS is requested, but forcing that the `Stream.null` is indeed the PTDS stream will leave no choice for the user to fallback to the legacy one if necessary for some reason, so perhaps it's good to still have that possibility regardless.

leofang replied:
@pentschev I just noticed in CuPy's test suite it's expected that when users do not specify any stream, this is true:
```python
assert cuda.Stream.null == cuda.get_current_stream()
```
See https://github.com/cupy/cupy/blob/master/tests/cupy_tests/cuda_tests/test_stream.py. I guess this assumption might have propagated to some user codes as well? At least I just recalled I got one 😅 But we don't plan to use PTDS so it's probably OK.

pentschev replied:
Hmm, that's interesting, indeed I didn't see that before. I suppose this could pose a problem, so we should probably try to disambiguate that first. I would still prefer to have a possibility to enforce using `Stream.null`, even when PTDS is enabled, but I'm happy to change that if the preferred is not to.

emcastillo replied:
Is using `Stream.null` when PTDS is enabled allowed?

leofang replied:
I think it is, but the semantics needs to be clearly defined (which is what we were discussing above -- should Stream.null mean stream 1 (legacy) or 2 (per-thread) when it's enabled?)

emcastillo replied:
We discussed and the majority was in favor of leaving `Stream.null` as the legacy one.
Anyway, we can just proceed as it is now and revisit it later.

pentschev replied:
Thanks for discussing that!

pentschev commented on cupy/cuda/stream.pyx:
```diff
@@ -275,9 +283,11 @@ class Stream(BaseStream):
+    def __init__(self, null=False, ptds=False, non_blocking=False):
         if null:
-            self.ptr = 0
+            self.ptr = runtime.cudaStreamDefault
```
Technically speaking, the `cudaStreamDefault`/`cudaStreamNonBlocking` are flags for [`cudaStreamCreateWithFlags`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1gb1e32aff9f59119e4d0a9858991c4ad3), but it happens that `cudaStreamDefault` is also `0`, which is consistent with the default behavior for when no streams are passed and is also used by many people, CuPy included. On the other hand `cudaStreamLegacy` and `cudaStreamPerThread` are _stream handles_ and these are the ones that we should be using when specifying a stream explicitly.

leofang replied:
Right, so this is exactly I was thinking: Don't we expect `self.ptr` to be a *stream handle* and so it should be `runtime. cudaStreamLegacy`?

pentschev replied:
_Probably_. The default was `0`, so I thought I would play it safe and avoid replacing that by `cudaStreamLegacy` for now. I believe the behavior should be the same though. Maybe it's a good time to replace and test that out, unless someone has any reasons not to do so that I may be overlooking.

pentschev replied:
I adjusted this now to `runtime.cudaStreamLegacy`, so far everything looks correct on my verification.

pentschev commented on cupy_backends/cuda/libs/cublas.pyx:
```diff
@@ -454,6 +454,8 @@ cdef _setStream(intptr_t handle):
     """
     if stream_module.enable_current_stream:
         setStream(handle, stream_module.get_current_stream_ptr())
+    elif stream_module.is_ptds_enabled():
+        setStream(handle, runtime.cudaStreamPerThread)
```
Are you suggesting we replace those two lines simply by the following?
```python
setStream(handle, stream_module.get_current_stream_ptr())
```

leofang replied:
Yes, but it's already done above.

pentschev replied:
It isn't, note how above there's `if stream_module.enable_current_stream`. The `enable_current_stream` is only set when using an explicit stream. However, we could perhaps remove that conditional too, I'm not sure if that would work though, my understanding is that `get_current_stream_ptr` returns `self.current_stream`, which is undefined until setting a stream, or am I missing something else?

leofang replied:
Ah, that I know. `self.current_stream` is a cdef attribute, so it's implicitly initialized to `0`. We should probably add a `__cinit__` method in `cupy_backends/cuda/stream.pyx` to ensure it's properly initialized (to whatever we eventually agree upon in https://github.com/cupy/cupy/pull/4322#discussion_r530601255. 

btw I would like to push a house keeping PR after this one is merged, so that everything in `cupy_backends/cuda/stream.pyx` is moved to `cupy/cuda/stream.pyx` (in the exactly opposite direction as proposed in https://github.com/cupy/cupy/pull/3584).

leofang replied:
For `stream_module.enable_current_stream`, I think @takagi is more knowledgable since he recently worked on the fix #4173. My understanding is `stream_module.enable_current_stream` is always set to `True` when a user attempt to use a stream (either as a context manager (`with stream:`) or explicitly turning it on (`stream.use()`).

However, I do think we should explicitly initialize it to `True`: note that it's a module attribute left uninitialized:
https://github.com/cupy/cupy/blob/d8f8a31393f5c7d3993cce951fcd29886b02c458/cupy_backends/cuda/stream.pxd#L3
So in a sense #4173 is a brute force fix (by overwriting it) 😅

pentschev replied:
> However, I do think we should explicitly initialize it to `True`: note that it's a module attribute left uninitialized:

So when would it be `False`? If we're forcing an initialization to `True`, it becomes a completely unnecessary attribute, as I do see it being set to `False` anywhere.

leofang replied:
I feel it is completely unnecessary to have.

pentschev replied:
I removed it, adjusted the `setStream` functions and `get_current_stream_ptr`. I'm not very happy with the condition in https://github.com/cupy/cupy/pull/4322/files#diff-c7ca9ae40b6014a65e9d4420604c0e83adb8aadd77011ce04fd2026d74f43a01R59-R60 , it doesn't look nice to compare to `<void*>0`, and I'm not sure what is the most appropriate/preferred/cleanest way to do that with Cython.

leofang replied:
`current_stream` could be changed to `intptr_t` IMHO so that we just compare with `0`. I don't have strong opinion on this, but would like to note this is done in one of the two `stream` modules, and best to be unified when they are merged.

pentschev replied:
I did this now in https://github.com/cupy/cupy/pull/4322/commits/ece5c5569d65633861e4c8b139e791cedb2e2b6b, seems that this change doesn't require modifications elsewhere.

jakirkham commented:
Thanks for sharing that Peter!

It's interesting that sometimes a gap shows up on one of the streams. Do we know why that is? No worries if not. Just curious 🙂

pentschev commented:
> It's interesting that sometimes a gap shows up on one of the streams. Do we know why that is? No worries if not. Just curious 🙂

The GPU has a limited amount of resources, most likely launching yet another concurrent kernel wouldn't fit due to one or multiple sets of resources (e.g., shared memory, registers, etc.).

leofang commented:
> > As of now, I also don't know how to write tests to ensure that we aren't using the default stream when setting `CUPY_CUDA_PER_THREAD_DEFAULT_STREAM=1`, something I verified with the help of `nsys` and `nvprof`.
> 
> btw This was my impression too, but I just begin to wonder if it's possible to add a callback to the stream, and let it throw out something that we can capture in Python? CuPy already wraps the stream callback API:
> https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.Stream.html#cupy.cuda.Stream.add_callback
> (Disclaimer: I've never tried this, I just read it from the Programming Guide 😅)

I ended up working on #4338 out of curiousity. Note this works:
```python
import cupy as cp


def callback(data):
    print('I am called!')
    data.append('catch me if you can')

def do_work(stream):
    data = []
    stream.launch_host_func(callback, data)
    with stream:
        a = cp.random.random(100)
        a *= 3
    assert data[0] == 'catch me if you can'
    assert len(data) == 1

s = cp.cuda.Stream()
do_work(s)
cp.cuda.Device().synchronize()
print('done')
```
I think the idea also applies if you change `stream.launch_host_func()` (to be added in #4338) to `stream.add_callback()`, with adjusting the `callback` function signature. So perhaps what we could do in the tests is to add callbacks to both legacy and per-thread default streams, check the returned data, and make sure only either of them is used but not both (depending on the test condition).

pentschev commented:
@leofang thanks for checking the `callback` idea, I'll try to write some tests using that for this PR.

leofang commented on cupy_backends/cuda/api/runtime.pxd:
```diff
@@ -524,8 +524,13 @@ cpdef enum:
 
-    streamDefault = 0
-    streamNonBlocking = 1
+    # cudaStream flags
+    cudaStreamDefault = 0
+    cudaStreamNonBlocking = 1
+
+    # cudaStream handles
+    cudaStreamLegacy = 1
+    cudaStreamPerThread = 2
```
Sorry, forgot to say that CuPy's convention for wrapping the CUDA API is to remove the prefix `cuda`. I think someone messed up in a few fields here and there (including me 😅), but this generally holds (see also the preamble in this file). 

In particular, `streamDefault` and `streamNonBlocking` have been around for a long time, and renaming them would cause backward incompatibility (they are public low-level API, and users could access them through `cupy.cuda.*`).

pentschev replied:
Got it, sorry for missing that, I reverted changes to the CuPy standard, and also changed new names to `streamLegacy`/`streamPerThread`.

leofang commented on cupy_backends/cuda/libs/cudnn.pyx:
```diff
@@ -6,6 +6,7 @@ cimport cython  # NOQA
 from libcpp cimport vector
 
 from cupy_backends.cuda.api cimport driver
+from cupy_backends.cuda.api cimport runtime
```
I think this import is no longer needed?

pentschev replied:
Good catch, I removed those now.

leofang reviewed (commented):
Thanks, @pentschev! LGTM except for a few minor points. I think we are left with
- Discussion on the `Stream.null` singleton
- Tests

pentschev commented on cupy/cuda/stream.pyx:
```diff
@@ -275,11 +285,14 @@ class Stream(BaseStream):
     """
 
-    def __init__(self, null=False, non_blocking=False):
+    def __init__(self, null=False, ptds=False, non_blocking=False):
```
Good point, this is done now.

pentschev commented:
@leofang  I extended existing tests to support PTDS as well, see https://github.com/cupy/cupy/pull/4322/commits/16650ee2c2bdb020088e0fa576cdcbc7a434e8d0 . However, `test_del` fails only for the PTDS test, I'm really not sure why that is, and honestly, I can't understand the logic behind it but it's reproducible on an interactive session:

```python
In [1]: import cupy

In [2]: stream = cupy.cuda.Stream(null=False, ptds=True).use()

In [3]: stream_ptr = stream.ptr

In [4]: del stream
Exception ignored in: <bound method Stream.__del__ of <cupy.cuda.stream.Stream object at 0x7f92f15eae50>>
Traceback (most recent call last):
  File "cupy/cuda/stream.pyx", line 308, in cupy.cuda.stream.Stream.__del__
  File "cupy_backends/cuda/api/runtime.pyx", line 784, in cupy_backends.cuda.api.runtime.streamDestroy
  File "cupy_backends/cuda/api/runtime.pyx", line 247, in cupy_backends.cuda.api.runtime.check_status
cupy_backends.cuda.api.runtime.CUDARuntimeError: cudaErrorInvalidResourceHandle: invalid resource handle

In [5]: assert cupy.cuda.Stream.ptds == cupy.cuda.get_current_stream()
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-5-d014826c55ae> in <module>
----> 1 assert cupy.cuda.Stream.ptds == cupy.cuda.get_current_stream()

AssertionError:
```

Any ideas what could be going wrong?

pentschev commented:
Nevermind, I just understood what I did wrong, after `del stream` CuPy should fallback to its default (legacy stream), so I shouldn't be comparing it to the stream that I just deleted, should be fixed in https://github.com/cupy/cupy/pull/4322/commits/5e0d863b5654b60612cbc4b3d445f825673f2af0 .

pentschev commented:
@leofang also note how I checked for the stream in the callback function, I'm not sure if that will still be possible with `cudaLaunchHostFunc` though.

leofang commented:
> Nevermind, I just understood what I did wrong, after `del stream` CuPy should fallback to its default (legacy stream), so I shouldn't be comparing it to the stream that I just deleted, should be fixed in [5e0d863](https://github.com/cupy/cupy/commit/5e0d863b5654b60612cbc4b3d445f825673f2af0) .

Good catch! Yeah deleting a stream could be surprising, see also #4337.

leofang commented:
Quick question for the purpose of #4357: If we now explicitly initialize to either `cudaStreamLegacy` or `cudaStreamPerThread`, I think it's no longer possible to have `stream.ptr == 0`, right?

pentschev commented:
> Quick question for the purpose of #4357: If we now explicitly initialize to either `cudaStreamLegacy` or `cudaStreamPerThread`, I think it's no longer possible to have `stream.ptr == 0`, right?

It isn't, unless the user forcefully does `stream.ptr = 0`. If the user doesn't, do you foresee any issues with not ever having `stream.ptr == 0`? I'm not sure if I'm perhaps overlooking something here.

leofang commented:
> > Quick question for the purpose of #4357: If we now explicitly initialize to either `cudaStreamLegacy` or `cudaStreamPerThread`, I think it's no longer possible to have `stream.ptr == 0`, right?
> 
> It isn't, unless the user forcefully does `stream.ptr = 0`. If the user doesn't, do you foresee any issues with not ever having `stream.ptr == 0`? I'm not sure if I'm perhaps overlooking something here.

One nice thing by excluding the possibility of `stream.ptr==0` is that we don't have to compare with `<void*>0` anymore (https://github.com/cupy/cupy/pull/4322#discussion_r531117864), so perhaps we could clean up a bit. If we really want to stay on the safe side, we could internally convert a `0` specified by the user to `cudaStreamLegacy`, making it fully backward compatible.

pentschev commented:
> One nice thing by excluding the possibility of `stream.ptr==0` is that we don't have to compare with `<void*>0` anymore ([#4322 (comment)](https://github.com/cupy/cupy/pull/4322#discussion_r531117864)), so perhaps we could clean up a bit. If we really want to stay on the safe side, we could internally convert a `0` specified by the user to `cudaStreamLegacy`, making it fully backward compatible.

Overall I like this idea. But if the user assigns `stream.ptr = 0`, when/where do you think would be appropriate to force a fallback to `cudaStreamLegacy`?

leofang commented:
> Overall I like this idea. But if the user assigns `stream.ptr = 0`, when/where do you think would be appropriate to force a fallback to `cudaStreamLegacy`?

Actually I don't see such a possibility (user sets `0`) with the current PR except when `cupy.cuda.ExternalStream` is used to wrap this value, am I missing something?! 

If desired, we could modify `get_current_stream_ptr()` in `cupy_backends/cuda/stream.pyx`, say emit a warning if `0` is provided and change it to `cudaStreamLegacy`. (I think we should wait for @kmaehashi before making any change.)

pentschev commented:
> > Overall I like this idea. But if the user assigns `stream.ptr = 0`, when/where do you think would be appropriate to force a fallback to `cudaStreamLegacy`?
> 
> Actually I don't see such a possibility (user sets `0`) with the current PR except when `cupy.cuda.ExternalStream` is used to wrap this value, am I missing something?!
> 
> If desired, we could modify `get_current_stream_ptr()` in `cupy_backends/cuda/stream.pyx`, say emit a warning if `0` is provided and change it to `cudaStreamLegacy`. (I think we should wait for @kmaehashi before making any change.)

What I mean is, what if the user decides to do something like below?

```python
In [1]: import cupy

In [2]: stream = cupy.cuda.Stream()

In [3]: stream.ptr
Out[3]: 94881341085904

In [4]: stream.ptr = 0

In [5]: stream.ptr
Out[5]: 0
```

I know this isn't generally expected, but given Python doesn't have private attributes, the user could very much do something like the above.

leofang commented:
> I know this isn't generally expected, but given Python doesn't have private attributes, the user could very much do something like the above.

I think there are a few places where a similar concern could be raised, but generally in Python we trust people to not do silly things and mess themselves around 🙂 If this is a serious concern, though, we can always decorate `ptr` with `@property`, or cythonize the `BaseStream` class so that `ptr` is marked `cdef readonly`, as is done in `cupy.RawKernel` for example. I'd rather not do/consider it in this PR, though. What do you think?

pentschev commented:
> I think there are a few places where a similar concern could be raised, but generally in Python we trust people to not do silly things and mess themselves around 🙂

Makes sense, because if there's one thing we can expect from users is for them _not_ to do any silly things. 🤣 

> If this is a serious concern, though, we can always decorate `ptr` with `@property`, or cythonize the `BaseStream` class so that `ptr` is marked `cdef readonly`, as is done in `cupy.RawKernel` for example. I'd rather not do/consider it in this PR, though. What do you think?

I think protecting `ptr` is generally a good idea, but I agree we can leave it for later.

pentschev commented:
I merged master into this branch and did some more testing on my side, everything looks good to me. It would be great if @emcastillo could take a look too! :)

emcastillo commented on cupy/cuda/stream.pyx:
```diff
@@ -303,11 +313,14 @@ class Stream(BaseStream):
     """
 
-    def __init__(self, null=False, non_blocking=False):
+    def __init__(self, null=False, non_blocking=False, ptds=False):
```
Let's add ptds to the docstring of `Stream`

pentschev replied:
Done in the latest commits, please note I did a few changes to clarify the order that arguments respect.

leofang commented on tests/cupy_tests/cuda_tests/test_stream.py:
```diff
@@ -6,21 +6,32 @@
 
-    def check_del(self, null):
-        stream = cuda.Stream(null=null).use()
+    def check_del(self, test_stream, null, ptds):
```
I don't think `test_stream` is used?

pentschev replied:
You're right, another leftover from my initial attempt at having this working. Should be fixed now. :)

pentschev commented on cupy/cuda/stream.pyx:
```diff
@@ -291,23 +301,33 @@ class Stream(BaseStream):
-    def __init__(self, null=False, non_blocking=False):
+    def __init__(self, null=False, non_blocking=False, ptds=False):
         if null:
-            self.ptr = 0
+            self.ptr = runtime.streamLegacy
```
Why? You could still set `my_stream.ptr = 0`, couldn't you? Or is there a specific use case you're referring to?

leofang replied:
I was merely leaving a note for other reviewers 😅 and I think it is actually a good thing that we kill stream 0. 
> You could still set `my_stream.ptr = 0`, couldn't you?

That is possible, true, but isn't something that is common. I meant if sticking to the exposed APIs stream 0 wouldn't be used.

leofang replied:
> In theory this eliminates the possibility of using stream 0 throughout the codebase.

This is reverted in 3cc00ce and is no longer valid; stream 0 is still used (by `Stream.null`).

pentschev commented on cupy/cuda/stream.pyx:
```diff
@@ -291,23 +301,33 @@ class Stream(BaseStream):
+        ptds (bool): If ``True`` and ``null`` is ``False``, the per-thread
+            default stream is used. Note that you can also use ``Stream.ptds``
+            singleton object instead of creating new per-thread default stream
+            object.
```
I'm not entirely sure this is the right place, perhaps it would fit best in https://github.com/cupy/cupy/blob/master/docs/source/reference/environment.rst ?

pentschev replied:
I've added documentation for the environment variable in https://github.com/cupy/cupy/pull/4322/commits/7acb9c4e48ccb3cc624f04d4eadd780c6a04d0fe

leofang commented:
Jenkins, test this please

leofang commented:
Failed tests are likely due to NumPy 1.20.

pentschev commented:
> Failed tests are likely due to NumPy 1.20.

Those are waiting for https://github.com/cupy/cupy/pull/4604, is that right?

leofang commented:
pfnCI, test this please

leofang commented:
@pentschev Could you check the NCCL failure? Maybe related to invalid stream pointer?

pentschev commented:
Certainly, thanks for the ping @leofang !

pentschev commented:
This seems to be either a bug or unsupported use case, I can also reproduce it with C++ code. I filed an internal bug ID 200699044 and will update here when I hear back from developers.

In the meantime, I think we'll need to discuss what's the best approach here. Both `cudaStreamLegacy` and `cudaStreamPerThread` fail in that test, which means we currently can't use any of them. Given we're replacing `Stream.null.ptr` with `cudaStreamLegacy`, that seems it will break any NCCL uses, but we could revert that to the null stream for the time being, any other ideas? It seems that the PTDS case won't work at all with NCCL at the moment, unfortunately.

leofang commented:
> It seems that the PTDS case won't work at all with NCCL at the moment, unfortunately.

Hi Peter, I did some manual tests and I think the "bug" is limited to when we initialize the communicator for all devices in the same process (via`ncclCommInitAll`). But if the initialization is done by `ncclCommInitRank` with multiple processes, which is the most common use case IMHO, then it happily accepts stream 1 & 2. You can try this code with MPI: 
```python
# mpirun -n 2 python test_nccl.py

from mpi4py import MPI
import cupy as cp


comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()
print(rank, size)

cp.cuda.Device(rank).use()

if rank == 0:
    n_id = cp.cuda.nccl.get_unique_id()
else:
    n_id = None
n_id = comm.bcast(n_id, root=0)
n_comm = cp.cuda.nccl.NcclCommunicator(size, n_id, rank)

a = cp.arange(10)
n_comm.allReduce(a.data.ptr, a.data.ptr,
                 10*a.dtype.itemsize, cp.cuda.nccl.NCCL_FLOAT64, cp.cuda.nccl.NCCL_SUM,
                 2)  # use stream 2
print(a)
```
Perhaps we could simply print a big warning when `initAll()` is called to tell users to not use stream 1 & 2 until it's fixed by upstream?

leofang commented:
(btw I found this as I was wondering if the bug happens only when `send`/`recv` is called, but it's not the case)

pentschev commented:
Thanks @leofang for further investigating this.

> Perhaps we could simply print a big warning when `initAll()` is called to tell users to not use stream 1 & 2 until it's fixed by upstream?

I would be fine with that, but the part that becomes maybe syntactically wrong then is that we moved the meaning of `Stream.null` to `cudaStreamLegacy` (stream 1). So in a sense, the user would claim "I'm using the null stream with `Stream.null`". Particularly, I have no objections to either emitting a warning or moving back to `Stream.null.ptr = 0` for the time being.

> (btw I found this as I was wondering if the bug happens only when `send`/`recv` is called, but it's not the case)

Indeed, I observed the same in C++ with `ReduceAll`.

emcastillo commented:
Yes, let's revert to the null stream.
Thanks!

leofang commented:
> Particularly, I have no objections to either emitting a warning or moving back to `Stream.null.ptr = 0` for the time being.

Indeed, it seems to be a good idea to revert to the status quo (stream 0 being null).

pentschev commented:
I reverted it back to `streamDefault`. I worked with NCCL developers and the issue has already been fixed, it was simply the case that special threads were just not checked for. That fix should be in the version after `2.8.3-1`, so I added a `TODO` note in the code to change that in the future.

pentschev commented:
pfnCI, test this please

leofang commented on cupy/cuda/stream.pyx:
```diff
@@ -291,23 +301,36 @@ class Stream(BaseStream):
+            # TODO(pentschev): move to streamLegacy. This wasn't possible
+            # because of a NCCL bug that should be fixed in the version
+            # following 2.8.3-1.
+            self.ptr = runtime.streamDefault
```
nitpick: `streamDefault` is a flag, not a stream macro
```suggestion
            self.ptr = 0
```

pentschev replied:
Stop using my own arguments from https://github.com/cupy/cupy/pull/4322#discussion_r530302943 against myself! 🤣

leofang replied:
Ah ha ha, forgot about that one! 😄

pentschev replied:
That's why you shouldn't write things that can be used against yourself, I trapped myself with that one! 😄

leofang commented:
Jenkins, test this please

pentschev commented:
I was informed that the fix for `cudaStreamLegacy`/`cudaStreamPerThread` should be in NCCL 2.9, which will be out probably in Q2 2021.

leofang reviewed (approved):
LGTM! Thanks @pentschev for pushing it through.

pentschev commented:
Thanks for help with reviews here @leofang and @emcastillo ! 😄
